File: plugin/src/test/java/com/oldstats/tracking/LootTrackerTest.java
package com.oldstats.tracking;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.oldstats.api.OldStatsApiClient;
import com.oldstats.api.StatEvent;
import java.time.Duration;
import java.util.Collections;
import java.util.List;
import net.runelite.api.ItemComposition;
import net.runelite.api.NPC;
import net.runelite.client.events.NpcLootReceived;
import net.runelite.client.game.ItemManager;
import net.runelite.client.game.ItemStack;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

public class LootTrackerTest {
    private OldStatsApiClient apiClient;
    private ItemManager itemManager;
    private LootTracker tracker;

    @Before
    public void setUp() {
        apiClient = mock(OldStatsApiClient.class);
        itemManager = mock(ItemManager.class);
        tracker = new LootTracker(apiClient, itemManager);
    }

    private NPC npc(String name, int id) {
        NPC n = mock(NPC.class);
        when(n.getName()).thenReturn(name);
        when(n.getId()).thenReturn(id);
        return n;
    }

    private void stubItem(int itemId, String name, int price) {
        ItemComposition composition = mock(ItemComposition.class);
        when(composition.getName()).thenReturn(name);
        when(itemManager.getItemComposition(itemId)).thenReturn(composition);
        when(itemManager.getItemPrice(itemId)).thenReturn(price);
    }

    @Test
    public void emitsAKillAndADropForEveryItemInTheLoot() {
        stubItem(21295, "Draconic visage", 8_500_000);
        tracker.onNpcLootReceived(new NpcLootReceived(npc("Vorkath", 8058), List.of(new ItemStack(21295, 1))));

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient, times(2)).enqueue(captor.capture());

        StatEvent.Kill kill = (StatEvent.Kill) captor.getAllValues().get(0);
        assertEquals("Vorkath", kill.npcName);
        assertEquals(Integer.valueOf(8058), kill.npcId);
        assertTrue(kill.isBoss);

        StatEvent.Drop drop = (StatEvent.Drop) captor.getAllValues().get(1);
        assertEquals("Draconic visage", drop.itemName);
        assertEquals(Integer.valueOf(21295), drop.itemId);
        assertEquals(1, drop.quantity);
        assertEquals(8_500_000L, drop.value);
    }

    @Test
    public void multipliesUnitPriceByQuantityForStackedDrops() {
        stubItem(12934, "Zulrah's scales", 20);
        tracker.onNpcLootReceived(new NpcLootReceived(npc("Zulrah", 2042), List.of(new ItemStack(12934, 1000))));

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient, times(2)).enqueue(captor.capture());
        StatEvent.Drop drop = (StatEvent.Drop) captor.getAllValues().get(1);
        assertEquals(20_000L, drop.value);
    }

    @Test
    public void flagsANonBossKillCorrectlyAndEmitsNoDropsWhenTheresNoLoot() {
        tracker.onNpcLootReceived(new NpcLootReceived(npc("Chicken", 41), Collections.emptyList()));

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient).enqueue(captor.capture());
        StatEvent.Kill kill = (StatEvent.Kill) captor.getValue();
        assertEquals("Chicken", kill.npcName);
        assertTrue(!kill.isBoss);
    }

    @Test
    public void fallsBackToUnknownWhenTheNpcHasNoName() {
        NPC unnamed = mock(NPC.class);
        when(unnamed.getId()).thenReturn(1);
        tracker.onNpcLootReceived(new NpcLootReceived(unnamed, Collections.emptyList()));

        ArgumentCaptor<StatEvent> captor = ArgumentCaptor.forClass(StatEvent.class);
        verify(apiClient).enqueue(captor.capture());
        assertEquals("Unknown", ((StatEvent.Kill) captor.getValue()).npcName);
    }

    @Test
    public void recentBossKillRemembersOnlyTheMostRecentBossWithinTheGivenWindow() {
        assertNull(tracker.recentBossKill(Duration.ofSeconds(10)));

        tracker.onNpcLootReceived(new NpcLootReceived(npc("Vorkath", 8058), Collections.emptyList()));
        assertEquals("Vorkath", tracker.recentBossKill(Duration.ofSeconds(10)));

        // A non-boss kill shouldn't clear or override the last boss kill.
        tracker.onNpcLootReceived(new NpcLootReceived(npc("Chicken", 41), Collections.emptyList()));
        assertEquals("Vorkath", tracker.recentBossKill(Duration.ofSeconds(10)));
    }

    @Test
    public void recentBossKillReturnsNullOnceTheWindowHasElapsed() {
        tracker.onNpcLootReceived(new NpcLootReceived(npc("Vorkath", 8058), Collections.emptyList()));
        assertNull(tracker.recentBossKill(Duration.ZERO.minusNanos(1)));
    }
}

File: plugin/src/test/kotlin/com/oldstats/tracking/LootTrackerTest.kt
package com.oldstats.tracking

import com.oldstats.api.OldStatsApiClient
import com.oldstats.api.StatEvent
import net.runelite.api.ItemComposition
import net.runelite.api.NPC
import net.runelite.client.events.NpcLootReceived
import net.runelite.client.game.ItemManager
import net.runelite.client.game.ItemStack
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.time.Duration
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class LootTrackerTest {
    private lateinit var apiClient: OldStatsApiClient
    private lateinit var itemManager: ItemManager
    private lateinit var tracker: LootTracker

    @Before
    fun setUp() {
        apiClient = mock()
        itemManager = mock()
        tracker = LootTracker(apiClient, itemManager)
    }

    private fun npc(name: String, id: Int): NPC {
        val n: NPC = mock()
        whenever(n.name).thenReturn(name)
        whenever(n.id).thenReturn(id)
        return n
    }

    private fun stubItem(itemId: Int, name: String, price: Int) {
        val composition: ItemComposition = mock()
        whenever(composition.name).thenReturn(name)
        whenever(itemManager.getItemComposition(itemId)).thenReturn(composition)
        whenever(itemManager.getItemPrice(itemId)).thenReturn(price)
    }

    @Test
    fun `emits a kill and a drop for every item in the loot`() {
        stubItem(21295, "Draconic visage", 8_500_000)
        tracker.onNpcLootReceived(NpcLootReceived(npc("Vorkath", 8058), listOf(ItemStack(21295, 1))))

        val captor = argumentCaptor<StatEvent>()
        verify(apiClient, org.mockito.kotlin.times(2)).enqueue(captor.capture())

        val kill = captor.allValues[0] as StatEvent.Kill
        assertEquals("Vorkath", kill.npcName)
        assertEquals(8058, kill.npcId)
        assertTrue(kill.isBoss)

        val drop = captor.allValues[1] as StatEvent.Drop
        assertEquals("Draconic visage", drop.itemName)
        assertEquals(21295, drop.itemId)
        assertEquals(1, drop.quantity)
        assertEquals(8_500_000L, drop.value)
    }

    @Test
    fun `multiplies unit price by quantity for stacked drops`() {
        stubItem(12934, "Zulrah's scales", 20)
        tracker.onNpcLootReceived(NpcLootReceived(npc("Zulrah", 2042), listOf(ItemStack(12934, 1000))))

        val captor = argumentCaptor<StatEvent>()
        verify(apiClient, org.mockito.kotlin.times(2)).enqueue(captor.capture())
        val drop = captor.allValues[1] as StatEvent.Drop
        assertEquals(20_000L, drop.value)
    }

    @Test
    fun `flags a non-boss kill correctly and emits no drops when there's no loot`() {
        tracker.onNpcLootReceived(NpcLootReceived(npc("Chicken", 41), emptyList()))

        val captor = argumentCaptor<StatEvent>()
        verify(apiClient).enqueue(captor.capture())
        val kill = captor.firstValue as StatEvent.Kill
        assertEquals("Chicken", kill.npcName)
        assertTrue(!kill.isBoss)
    }

    @Test
    fun `falls back to Unknown when the npc has no name`() {
        val unnamed: NPC = mock()
        whenever(unnamed.id).thenReturn(1)
        tracker.onNpcLootReceived(NpcLootReceived(unnamed, emptyList()))

        val captor = argumentCaptor<StatEvent>()
        verify(apiClient).enqueue(captor.capture())
        assertEquals("Unknown", (captor.firstValue as StatEvent.Kill).npcName)
    }

    @Test
    fun `recentBossKill remembers only the most recent boss, within the given window`() {
        assertNull(tracker.recentBossKill(Duration.ofSeconds(10)))

        tracker.onNpcLootReceived(NpcLootReceived(npc("Vorkath", 8058), emptyList()))
        assertEquals("Vorkath", tracker.recentBossKill(Duration.ofSeconds(10)))

        // A non-boss kill shouldn't clear or override the last boss kill.
        tracker.onNpcLootReceived(NpcLootReceived(npc("Chicken", 41), emptyList()))
        assertEquals("Vorkath", tracker.recentBossKill(Duration.ofSeconds(10)))
    }

    @Test
    fun `recentBossKill returns null once the window has elapsed`() {
        tracker.onNpcLootReceived(NpcLootReceived(npc("Vorkath", 8058), emptyList()))
        assertNull(tracker.recentBossKill(Duration.ZERO.minusNanos(1)))
    }
}
